File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int[] res = new int[len];
        res[0] = nums[0];
        for(int i = 1; i < len; i++){
            res[i] = res[i-1] * nums[i];
        }
        int product = nums[len - 1];
        res[len - 1] = res[len - 2];
        for(int i= len - 2; i > 0;  i--){
            res[i] = res[i-1] * product;
            product *= nums[i];
        }
        res[0] = product;
        return res;
    }
}

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        length = len(nums)
        res = [0]*length
        res[0] = nums[0]
        for i in range(1,length):
            res[i] = res[i-1] * nums[i]
        product = nums[length - 1]
        res[length-1] = res[length-2]
        for i in range(length-2,0,-1):
            res[i] = res[i-1] * product
            product *= nums[i] 
        res[0] = product
        return res